### ISSUE TITLE
Add ability to control logger output level

### ISSUE THREAD

**Issue by havebeenfitz** [tuist/tuist#3191]
## Context 🕵️‍♀️

> It would be great to have logger levels for `tuist generate`, `tuist dependencies fetch/update` and maybe some other commands for output to be concise if needed, for example:

``` bash
Compiling assets... Success!
Generating .xcodeproj files... Success!
...
Installing CocoaPods... Success!
```

## Proposal 🌱

> Maybe add some parameters to main commands to control the logger output, for example `--logs` with possible set of values like 
- `stages` 
- `errors-only` 
- `warnings` 
- `verbose`

Value names are just for example)

**Comment by fortmarek** [comment]
Hi @havebeenfitz!

It is _possible_ to let users customize the logging behavior further (that is apart from using `--verbose` flag). Under the hood, we use [swift-log](https://github.com/apple/swift-log) package.

So e.g. if you wanted to have `errors-only`, we could do `LoggingSystem.bootstrap(StreamLogHandler.standardError)`, so that `stderr` becomes standard output.

There are also a [bunch of levels](https://github.com/apple/swift-log/blob/main/Sources/Logging/Logging.swift#L737) on which basis we could filter what should be logged.

From other tools that I have used, I can say that having a `--silent` option is quite usual, so adding that one makes probably the most sense for me - however, I'm not sure how much value it brings if we just log _nothing_. But e.g. cocoapods has this approach. Another tool yarn also has `--silent` flag but they still print output from scripts, so it was rather to enable piping the output of scripts while disregarding the rest of what yarn outputs.

All-in-all, I'm not against adding new flags controlling logger behavior as long as they adhere to some convention already set by other tools. Logging is so universal that we should not reinvent the wheel here. 

That being said, I think the best solution would be if we could maybe come up with a list of outputs that users don't find useful and either delete those or change their behavior. For example, if you have a lot of projects, the output contains a lot of messages like `Generating project XYZ...` - what we could do instead is always replace that line, similarly to what happens when you run `swift build` and see the number of steps. This route is the most difficult one as it means reevaluating a lot of output we do but ultimately, it will help the final user experience the most IMO.

**Comment by havebeenfitz** [comment]
Hi @fortmarek, sorry for the late response, I was on a quick vacation)

Replacing the line seems like the best solution, I agree) However, the `--silent` flag would be useful too in my opinion since a lot of teams use their own script wrappers. At least for us at Vivid it would be great to have that and we could fallback to default logging and even verbose if anything goes wrong

**Comment by fortmarek** [comment]
> Replacing the line seems like the best solution, I agree

Indeed! I can give you some pointers if you'd like to implement this, I have done something similar in the past.

> However, the --silent flag would be useful too in my opinion since a lot of teams use their own script wrappers.

I agree, it probably makes sense to add this. I'd be reluctant to adding additional flags to control the output, however.

**Comment by havebeenfitz** [comment]
> I can give you some pointers if you'd like to implement this

Sure, I would love to do this and some initial help would be nice)

So the plan is to replace the line for default behavior and completely silence the output for `--silent` option, right?

**Comment by fortmarek** [comment]
> So the plan is to replace the line for default behavior and completely silence the output for --silent option, right?

Yes, I think that would be the way to go. I'd also make sure that the line is not replaced when `--verbose` is run.

> Sure, I would love to do this and some initial help would be nice)

There's [this](https://github.com/apple/swift-tools-support-core/blob/435a2708a6e486d69ea7d7aaa3f4ad243bc3b408/Sources/TSCUtility/ProgressAnimation.swift#L32) that you can use to achieve it - you supply it with the number of steps (the number of projects we need to generate) and at which stage we are. For each step you can also update the text (that should be `Generating xyz...`).

I've actually had a talk on this topic at [appbuilders](https://www.youtube.com/watch?v=P_Ld_2U6-KE) last year. The example code is [here](https://github.com/fortmarek/appbuilders).

**Comment by havebeenfitz** [comment]
Thanks! I can't commit to a certain date, but will try to implement this in near future)

**Comment by fortmarek** [comment]
We never expect someone to commit to a certain date as our work is done in our free time - feel free to take as much time as you need!

**Comment by dxmvsh** [comment]
Hi, @fortmarek 

Is it still relevant to resolve this issue? Long time has passed, wanted to know if it is okay to take a chance to look into this. 

By the way, I looked at resources that you provided, especially very thankful for video you did at [appbuilders](https://www.youtube.com/watch?v=P_Ld_2U6-KE).
